fs/fs_zenfs.cc

```diff
@@ -491,6 +511,12 @@ IOStatus ZenFS::DeleteFile(std::string fname) {
     if (zoneFile != nullptr) {
       std::string record;
 
+      if (zoneFile->GetNrLinks()) {
```

Comment:
> I think it would be a better approach to _not_ distinguish between regular files and hard links. Instead, create files with just 1 hard link by default, and add more when the link API is invoked. That would reduce code complexity in a number of places, since we do not have to consider whether a file is a regular file or a link.

We are not distinguishing between regular links and hard links, the links also are listed by files_ list as individual files, but the same ZoneFile object backs all the hard linked files. So that any data/metadata update to either the source file or hard links is reflected in all the files.  This is far more convenient than maintaining each hard link as separate file with it's own ZoneFile object , as this would mean updating all linked files whenever the source file data/metadata changes, this also would require to maintain some kind of a 2-way link from source file to linked files to update changes. 


> For readability

Ok

Reply:
> We are not distinguishing between regular links and hard links, the links also are listed by files_ list as individual files, but the same ZoneFile object backs all the hard linked files. So that any data/metadata update to either the source file or hard links is reflected in all the files. This is far more convenient than maintaining each hard link as separate file with it's own ZoneFile object , as this would mean updating all linked files whenever the source file data/metadata changes, this also would require to maintain some kind of a 2-way link from source file to linked files to update changes.

I did not mean having a `ZoneFile` object for each link. Just, when we allocate the `ZoneFile`, insert the first path of the file into the `ZoneFile->linkfiles_`, so that `ZoneFile->GetNrLinks() >=1` is invariant. If we do that, we can simplify the code in a few places.

Reply:
I am not sure if it simplifies more. If it is part of linkfiles_ then we would need to ensure we don't try to re-insert it into 
files_ map, during decoding/encoding the updates.  It would add more corner cases in my opinion.

Reply:
This is the decode update now, from [here](https://github.com/westerndigitalcorporation/zenfs/pull/156/commits/7ff22e5200ee5c98158bd2a2e9b816c4b4a8ff76#diff-b12ba0b7fd2feee53dba42938e4031b32be21c71dc948ba759844c4c49be0a93R915-R941):
```c++
/* Remove all links from files_, get the updated linkfiles */
      for (uint32_t i = 0; i < zFile->GetNrLinks(); i++) {
        if (files_.find(zFile->GetLinkFilename(i)) != files_.end())
          files_.erase(zFile->GetLinkFilename(i));
        else
          return Status::Corruption("DecodeFileUpdateFrom: missing link file");
      }
      s = zFile->MergeUpdate(update, replace);
      update.reset();

      if (!s.ok()) return s;

      if (zFile->GetFilename() != oldName) {
        files_.erase(oldName);
        files_.insert(std::make_pair(zFile->GetFilename(), zFile));
      }

      for (uint32_t i = 0; i < zFile->GetNrLinks(); i++) {
        std::shared_ptr<ZoneFile> z = zFile;
        files_.insert(std::make_pair(zFile->GetLinkFilename(i), z));
      }
```
That could be changed to
```c++
/* Remove all links from files_, get the updated linkfiles */
      for (uint32_t i = 0; i < zFile->GetNrLinks(); i++) {
        if (files_.find(zFile->GetLinkFilename(i)) != files_.end())
          files_.erase(zFile->GetLinkFilename(i));
        else
          return Status::Corruption("DecodeFileUpdateFrom: missing link file");
      }
      s = zFile->MergeUpdate(update, replace);
      update.reset();

      if (!s.ok()) return s;

      for (uint32_t i = 0; i < zFile->GetNrLinks(); i++) {
        std::shared_ptr<ZoneFile> z = zFile;
        files_.insert(std::make_pair(zFile->GetLinkFilename(i), z));
      }
```
The first loop removes all links including the first _true_ name because that one is not special. The second loop would insert all links, including the first _true_ name. I don't see how that is more complex.

In the same way, we can completely drop `kFileName` from the on disk format, so encode/decode on disk format would be more simple.

Reply:
Yeah, this is another way of implementing the api.
Basically we remove the kFileName encoding and decoding and treat filenames as same as links do the filename encoding/decoding as part of linkfiles_ encoding and decoding. 

Pros is that it helps reducing few lines of code. 
Cons are that the filename_ processing also happens a part of linkfiles_ and it will not be straight forward for someone new looking at the code. 
I would still prefer keep them seperate, but we can discuss and take a final call.